src/env.cc

```diff
@@ -834,7 +836,8 @@ Environment::~Environment() {
   // FreeEnvironment() should have set this.
   CHECK(is_stopping());
 
-  if (options_->heap_snapshot_near_heap_limit > heap_limit_snapshot_taken_) {
+  if (heapsnapshot_near_heap_limit_callback_added()) {
+    set_heapsnapshot_near_heap_limit_callback_added(false);
```

Comment:
To avoid mistakes, we should probably create helper functions in `Environment` that adds/removes the callback as well as updating this field (which should also DCHECK the field)